stew/result.nim

```diff
@@ -232,20 +237,25 @@ template `and`*(self: Result, other: untyped): untyped =
 template `or`*(self: Result, other: untyped): untyped =
   ## Evaluate `other` iff not self.isOk, else return self
   ## fail-fast - will not evaluate other if a is a value
+  ##
+  ## TODO: This API is unsafe due to potential multiple
+  ## evaluation of the `self` parameter.
   if self.isOk: self
   else: other
 
-template catch*(body: typed): Result[type(body), ref Exception] =
+template catch*(body: typed): Result[type(body), ref CatchableError] =
   ## Convert a try expression into a Result
-  type R = Result[type(body), ref Exception]
+  type R = Result[type(body), ref CatchableError]
 
   try:
     R.ok(body)
-  except:
-    R.err(getCurrentException())
+  except CatchableError as e:
+    R.err(e)
 
-template capture*(T: type, e: ref Exception): Result[T, ref Exception] =
-  type R = Result[T, ref Exception]
+template capture*(T: type, someExceptionExpr: ref Exception): auto =
+  type ExceptionType = type(someExceptionExpr)
```

Comment:
This code may be a bit tricky to understand. 

The main idea is to preserve the type of the input exception. Since this is a template, even though the signatures says `ref Exception`, we are still able to obtain the precise type here, so we can return a `Result` with the correct type.

The `let e = someExceptionExpr` line is needed to avoid the double evaluation of the input parameter.

Reply:
right, good point - I also forgot to document this one it seems.. 

where would the double evaluation happen? I only see one evaluation, unless you count `typeof`?

Reply:
Ah, right, I first tried to refactor the code to use `system.getStackTraceEntries` instead of the `raise` trick and this needed the extra variable, but ultimately, it didn't work out and I restored the original version.